### DISCUSSION TITLE
Backing up file system for flashing

### DISCUSSION THREAD

**Discussion by ma261065** [micropython/micropython#12223] [ESP32] [answered]
Hi all,

I'm putting together instructions to a fab house to flash the boards that they are making for us.

I want to include our application files, but I don't want to "freeze" them in a firmware image, as I just want to use a stock image.

Is it "supported" to do the following steps to achieve this:

1) `import esp32`
`esp32.Partition.find(type=esp32.Partition.TYPE_DATA)`

This will list the data partitions. The filesystem is stored in the label=vfs

Make note of the address and size of the vfs partition (mine were address=5177344, size=11599872) and convert them to hex (address=0x4f0000, size=0xb10000 in my case)

2) Make a backup of the file system using:  `esptool -p COM4 -b 460800 --before default_reset --after no_reset --chip esp32s3  read_flash 0x4f0000 0xb10000 fsbackup.bin`

3) Then when they flash the chip from scratch, they can just include that file. e.g.

`esptool -p COM4 -b 460800 --before default_reset --after no_reset --chip esp32s3  write_flash --flash_mode dio --flash_size detect --flash_freq 80m 0x0 bootloader.bin 0x8000 partition-table.bin 0x10000 micropython.bin 0x4f0000 fsbackup.bin`

Also, is there any need to back-up and restore any of the other data partitions? (i.e. nvs, otadata, phy_init)

**Comment by jimmo** [comment, marked as answer]
@ma261065 Yes this is definitely supported.

FWIW, I'm guessing you're using a 16MiB-OTA configuration, in which case these numbers come from the partiton table: https://github.com/micropython/micropython/blob/master/ports/esp32/partitions-16MiB-ota.csv

(you can also extract this out of partition-table.bin, which is generated from that csv file)

For a given flash size + configuration, these offsets and sizes must remain fix, otherwise new firmware updates will be unable to access the filesystem partition.

For the other partitions, only `nvs` if you're using NVS (and want to restore that data).

**Reply by ma261065** [reply]
Thanks Jim. Yes, I'm using the 16MiB-OTA configuration.

I don't know what the NVS partition is used for, so I don't know whether I need to back it up or not...

**Reply by jimmo** [reply]
@ma261065 http://docs.micropython.org/en/latest/library/esp32.html#non-volatile-storage

**Comment by ma261065** [comment]
Got it, thanks. I wasn't sure if it was used by MicroPython itself for anything, but it looks like it's just a data store for the application to store key/value pairs in, which I'm not using.

All clear now - appreciate the response.
